src/main/java/Application/OpenCVClient/OpenCVClientOperator.java

```diff
@@ -0,0 +1,269 @@
+import Application.Utilities.ImageProcessor;
+import Application.Utilities.OpenCVUtil;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import org.opencv.core.Mat;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+
+import java.util.concurrent.Semaphore;
+
+public class OpenCVClientOperator {
+    private Socket clientSocket;
+    private BufferedReader in;
+    private DataOutputStream out;
+    private DetectMarker detector;
+    private ObjectMapper mapper;
+
+    private String fileName = "singlemarkerssource.png";
+    private String resultFileName = "detected.png";
+    private String fileExtension = "png";
+    private Mat subject;
+    private BufferedImage currentImage;
+
+    public Semaphore ipLock;
+    public Semaphore serverLock;
+
+    //Server Communication Properties
+    private int utilizeServer;
+    private String ip;
+    private int port;
+
+    public HeartBeatRunner beatRunner;
+    public MasterCommunicationRunner masterCommunicationRunner;
+    public ProcessingRunner processingRunner;
+
+    private static OpenCVClientOperator instance = null;
+
+    public OpenCVClientOperator() {
+        mapper = new ObjectMapper();
+        ipLock = new Semaphore(1);
+        serverLock = new Semaphore(1);
+        utilizeServer = 0;
+    }
+
+    public static OpenCVClientOperator getInstance() {
+        if (instance == null) {
+            instance = new OpenCVClientOperator();
+        }
+        return instance;
+    }
+
+    public void startConnection() {
+        try {
+            ipLock.acquire();
+            clientSocket = new Socket(ip, port);
+            out = new DataOutputStream(clientSocket.getOutputStream());
+            in = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
+        } catch (Exception e) {
+            e.printStackTrace();
+        } finally {
+            ipLock.release();
+        }
+    }
+
+    public String sendImage() throws IOException {
+        String resp = "";
+        Mat ids;
+        currentImage = ImageIO.read(new File(fileName));
+        ImageIO.write(currentImage, fileExtension, clientSocket.getOutputStream());
+        System.out.println("sendImage3");
+        resp = in.readLine();
+
+        return resp;
+    }
+
+    public void stopConnection() throws IOException {
+        in.close();
+        out.close();
+        clientSocket.close();
+    }
+
+    public String getFileName() {
+        return fileName;
+    }
+
+    public int useServer() {
+        return utilizeServer;
+    }
+
+    public void setServerUtilization(int input) {
+        try {
+            serverLock.acquire();
+            utilizeServer = input;
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        } catch (Exception e) {
+            e.printStackTrace();
```

Comment:
I see, I always added the general Exception in the end in case there's anything that I might have missed, but since there is only one specific, I agree that it's more confusing to add a general exception here

as for the setServerUtilization to false, I don't think it's wise since the Exception happened when acquiring the lock, so if we modify the value without ensuring that we were able to secure the lock to do so, I think it kinda defeats the purpose
that being said I'm not too familiar with distributed systems so I would like to hear your opinion on this as well

Reply:
You are right about modifying the value, it beats the purpose of the lock.

Now that you mention, I wanted to ask you about the locks. If the only thing that uses this resources is a singleton, shouldn't the lock be on the singleton instance itself, not inside? 

The locks are added to ensure that only one thread can access the resource (in this case `utilizeRunner`) at the same time, and that way avoid two threads working with different values. But if the resource is modified by a singleton, by definition, there would be only one instance accessing the resources (independent from the number of threads). However, the singleton itself becomes the shared resource. So I think it would make more sense to protect it than the resources. Do you see my point? Do you agree?